### ISSUE TITLE
[2.x OSS]: tags not disappearing after deleting by predicate

### ISSUE THREAD

**Issue by lephisto** [influxdata/influxdb#20704]
__Steps to reproduce:__
List the minimal actions needed to reproduce the behavior.

1. Have data with tags in a measurement.
2. Delete Data with a specific tag value:

```
influx delete --bucket "telegraf/autogen" --predicate '_measurement="devax_forecast" AND fcdate="fc_21012021"' -o NETZConsult --start '1970-01-01T00:00:00Z' --stop '2025-12-31T23:59:00Z')
```

3. Check of Tag value is removed from the DB:

```
show tag values from devax_forecast WITH key="fcdate"
```

and it's still showing up:

```
key    |value
fcdate |fc_01022021
fcdate |fc_02022021
fcdate |fc_03022021
fcdate |fc_04022021
fcdate |fc_21012021
fcdate |fc_22012021
fcdate |fc_23012021
fcdate |fc_24012021
fcdate |fc_25012021
fcdate |fc_26012021
fcdate |fc_27012021
fcdate |fc_27012021_erst
fcdate |fc_28012021
fcdate |fc_29012021
fcdate |fc_30012021
```

__Expected behavior:__
The tag should disappear.

__Actual behavior:__
The tag is still present, even if no records are assigned that tag.

__Environment info:__

* System info: Linux 4.4.0-198-generic x86_64
* InfluxDB version: InfluxDB 2.0.3 (git: fe04d346df) build_date: 2020-12-15T01:00:16Z
* Other relevant environment details: Runs from debian Packages on a Ubuntu 16.04.7

**Comment by lephisto** [comment]
Update: Issue still present in Influx 2.0.4 OSS

**Comment by nickwbaker** [comment]
This issue is affecting our team as well and we cannot delete data using the delete with predicate option. This is becoming a show stopper for us. The documentation indicates that this should work but it does not:

We are trying something simple like:
influx delete --bucket example-bucket \
  --start '1970-01-01T00:00:00Z' \
  --stop '2021-03-03T00:00:00Z' \
  --predicate 'DataSetId="78b6fdef-6908-4f26-85a2-e61be2f91a6d" AND _field="foo"'

The data is not deleted. 

If we remove the _field="foo" then it works to remove the data tagged with DataSetId="78b6fdef-6908-4f26-85a2-e61be2f91a6d". This was tested and verified by our dev teams as broken in 2.0.3, 2.0.4. 

So we can't delete data, which is killing us. Can anyone on influx team give an update on this. It seems to be an issue that has been around for a while an NOT addressed properly. Documentation is not accurate.

**Comment by danxmoran** [comment]
@nickwbaker deleting by `_field` is a separate issue, see #20399

**Comment by russorat** [comment]
Added https://github.com/influxdata/docs-v2/issues/2571 to update the documentation to include this.

**Comment by russorat** [comment]
@nickwbaker the way the influxdb storage engine works, deleting by a specific field is not possible. 

As for a workaround, you will need to write only the data you need to a new bucket and drop the old one.

```
from(bucket: "old_bucket") 
  |> range(start: -100y) 
  |> filter(fn: (r) =>  _field != "foo") 
  |> to(bucket: "new_bucket")
```
you can also do it from the command line by exporting the data from the old bucket as LP using the https://docs.influxdata.com/influxdb/v2.0/reference/cli/influxd/inspect/export-lp/ command and then ingesting that into a new bucket.

I know it's not the answer you wanted, but hopefully this can unblock you in the short term.

**Comment by lephisto** [comment]
@russorat I just want to clarify that the comment of @nickwbaker isn't exactly what I am experiencing. I experience the bug, that tag values still show up even if all data that might ever have been assigned to it has been deleted.

**Comment by MajorDallas** [comment]
Also having this issue. Version info:

```
# influx version
Influx CLI 2.0.3 (git: fe04d346df) build_date: 2020-12-15T01:00:16Z
```

I have a tag with two keys that need to be dropped as they contain invalid data that is interfering with other queries, but I cannot. I'm nearly at the point of running `sed` on the db files to get rid of them!

**Comment by midzelis** [comment]
I was able to delete tags using a similar command…

```shell
influx delete --bucket "ups" --predicate '_field="watts"' -o Primary --start "1970-01-01T00:00:00Z" --stop "2025-12-31T23:59:00Z"
```
where ‘watts’ is the name of the tag…

**Comment by 22ha** [comment]
I ran into the same problem. I used "drop series from 'measurement' where 'tag_key'='tag_value'". I dropped the series, but the tag values were still there.

**Comment by bygmakavi** [comment]
Anyone manage to fix this ? 
@lephisto 

I have the issue on 2.3.0

**Comment by ValkyrieOps** [comment]
Reporting this is still an issue on OSS 2.6.0

**Comment by peycho** [comment]
influxdb2 2.7.0-1
Still an issue.

**Comment by mattjudson** [comment]
I too appear to be struggling with this.... Deleted this data successfull but the tag still exists

root@8ab339126cfd:/# influx delete --bucket telegraf -o PBCSD -t ****   --start '1970-01-01T00:00:00Z'   --stop $(date +"%Y-%m-%dT%H:%M:%SZ")   --predicate '_measurement="an-hosts" AND "system-name"="drac-esxfhtst10"'

![image](https://github.com/influxdata/influxdb/assets/40123833/278f1517-b8d4-401e-8b13-c40954f38306)

**Comment by MattBrth** [comment]
Same here please fix this ...

**Comment by basalisco** [comment]
OSS 2.7.5 It continues to be problem. The tag is never deleted

**Comment by Tulux** [comment]
OSS edition don't seem really supported, might consider using another engine

**Comment by sanderson** [comment]
When deleting tags, the points with those tags are removed, but the tags themselves still exist in the index. Try rebuilding the index with [`influxd inspect build-tsi`](https://docs.influxdata.com/influxdb/v2/reference/cli/influxd/inspect/build-tsi/).

**Side note**: You shouldn't use the Data Explorer to test for the existence of a tag after you delete it. There are multiple forms of caching (database index, browser cache, etc.) that could make deleted tags still show up. If this is your only way to check, then be sure to do a hard refresh after deleting the tag and rebuilding the index.

**Comment by Tulux** [comment]
> When deleting tags, the points with those tags are removed, but the tags themselves still exist in the index. Try rebuilding the index with [`influxd inspect build-tsi`](https://docs.influxdata.com/influxdb/v2/reference/cli/influxd/inspect/build-tsi/).
> 
> **Side note**: You shouldn't use the Data Explorer to test for the existence of a tag after you delete it. There are multiple forms of caching (database index, browser cache, etc.) that could make deleted tags still show up. If this is your only way to check, then be sure to do a hard refresh after deleting the tag and rebuilding the index.

Thank you for your reply, indeed it works as expected.

It may worth linking https://docs.influxdata.com/influxdb/v2/admin/internals/tsi/rebuild-index/ into https://docs.influxdata.com/influxdb/v2/write-data/delete-data/ so that other users don't get confused.

**Comment by peycho** [comment]
Stopping the influxd service just to remove some tags is wild.
My setup has like 6k clients/connections and more than 2TB of data. Stopping the service "just because" is not acceptable option and I doubt it's only me.

There should be way cleaner and safer way of deleting tags.

**Comment by LorenzoGentili1** [comment]
Good afternoon everyone. I'm still experiencing the same behavior you have pointed out, no possibility to remove the tags after deleting the _predicate. Is there any turnaround for this issue? I'm currently running influxdb 2.7.

**Comment by philjb** [comment]
hi @LorenzoGentili1 -- Please [note the comment from Scott above](https://github.com/influxdata/influxdb/issues/20704#issuecomment-2299319872). There is the possibility to remove deleted tags through rebuilding the index in an offline process. You may use `influxd inspect build-tsi` but if you've deleted the index and series files [per the instructions](https://docs.influxdata.com/influxdb/v2/admin/internals/tsi/rebuild-index/), influxd will rebuild them on restart too (no need to use build-tsi in that case). When doing a version upgrade may be a good time to rebuild the index information since you are restarting influxd for the new version.

With our Enterprise 1 product, it's possible to do a rolling restart to rebuild indices without cluster downtime. That's not possible in single node OSS 2.x/1.x.

**Comment by davidby-influx** [comment]
Answered.
